packages/tre/src/plugins/d1/gateway.ts

```diff
@@ -0,0 +1,226 @@
+  };
+}
+function err(error: unknown): D1ErrorResponse {
+  return {
+    error: String(error),
+    success: false,
+    served_by,
+  };
+}
+
+export class D1Error extends HttpError {
+  constructor(cause: unknown) {
+    super(500, undefined, cause as Error);
+  }
+
+  toResponse(): Response {
+    return Response.json(err(this.cause));
+  }
+}
+
+type QueryRunner = (query: D1SingleQuery) => D1SuccessResponse;
+
+function normaliseParams(params: D1SingleQuery["params"]): any[] {
+  return (params ?? []).map((param) =>
+    // If `param` is an array, assume it's a byte array
+    Array.isArray(param) ? new Uint8Array(param) : param
+  );
+}
+function normaliseResults(rows: any[]): any[] {
+  return rows.map((row) =>
+    Object.fromEntries(
+      Object.entries(row).map(([key, value]) => [
+        key,
+        // If `value` is an array, convert it to a regular numeric array
+        value instanceof Buffer ? Array.from(value) : value,
+      ])
+    )
+  );
+}
+
+const DOESNT_RETURN_DATA_MESSAGE =
+  "The columns() method is only for statements that return data";
+const EXECUTE_RETURNS_DATA_MESSAGE =
+  "SQL execute error: Execute returned results - did you mean to call query?";
+function returnsData(stmt: Statement): boolean {
+  try {
+    stmt.columns();
+    return true;
+  } catch (e) {
+    // `columns()` fails on statements that don't return data
+    if (e instanceof TypeError && e.message === DOESNT_RETURN_DATA_MESSAGE) {
+      return false;
+    }
+    throw e;
+  }
+}
+
+export class D1Gateway {
+  private readonly db: DatabaseType;
+
+  constructor(private readonly log: Log, private readonly storage: Storage) {
+    this.db = storage.getSqliteDatabase();
+  }
+
+  #query: QueryRunner = (query) => {
+    const meta: OkMeta = { start: performance.now() };
+    // D1 only respects the first statement
+    const sql = splitSqlQuery(query.sql)[0];
+    const stmt = this.db.prepare(sql);
+    const params = normaliseParams(query.params);
+    let results: any[];
+    if (returnsData(stmt)) {
+      results = stmt.all(params);
+    } else {
+      // `/query` does support queries that don't return data,
+      // returning `[]` instead of `null`
+      const result = stmt.run(params);
+      results = [];
+      meta.last_row_id = Number(result.lastInsertRowid);
+      meta.changes = result.changes;
+    }
+    return ok(normaliseResults(results), meta);
+  };
+
+  #execute: QueryRunner = (query) => {
+    const meta: OkMeta = { start: performance.now() };
+    // D1 only respects the first statement
+    const sql = splitSqlQuery(query.sql)[0];
+    const stmt = this.db.prepare(sql);
+    // `/execute` only supports queries that don't return data
+    if (returnsData(stmt)) throw new Error(EXECUTE_RETURNS_DATA_MESSAGE);
+    const params = normaliseParams(query.params);
+    const result = stmt.run(params);
+    meta.last_row_id = Number(result.lastInsertRowid);
+    meta.changes = result.changes;
+    return ok(null, meta);
```

Comment:
Wait, so `execute` returns `null`, while `query` returns an empty array for statements that don't return data? That seems like a confusing API design to me (cc @rozenmd—what's the reasoning for this?)

Reply:
@geelen / @xtuc  would know the exact details, but we seem to only use `query`?